File: 剑指Offer27-er-cha-shu-de-jing-xiang-lcof/Offer27.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

/**
    Definition for a binary tree node.
**/
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode* mirrorTree(TreeNode* root) {
        if (root == nullptr) return root;
        TreeNode* tmp = root->left;
        root->left = root->right;
        root->right = tmp;
        mirrorTree(root->right);
        mirrorTree(root->left);
        return root;
    }
};

void print(TreeNode* t) {
    if (t == nullptr) return;
    queue<TreeNode*> q;
    q.push(t);
    int size = 0;
    cout << "[ ";
    while (!q.empty()) {
        size = q.size();
        for (int i = 0; i < size; ++i) {
            cout << q.front()->val << ", ";
            TreeNode* tmp = q.front();
            if (tmp->left) {
                q.push(tmp->left);
            }
            if (tmp->right) {
                q.push(tmp->right);
            }
            q.pop();
        }
    }
    cout << "]" << endl;
}

int main(int argc, char** args) {
    TreeNode* t1 = new TreeNode(4);
    TreeNode* t2 = new TreeNode(2);
    TreeNode* t3 = new TreeNode(7);
    TreeNode* t4 = new TreeNode(1);
    TreeNode* t5 = new TreeNode(3);
    TreeNode* t6 = new TreeNode(6);
    TreeNode* t7 = new TreeNode(9);

    t1->left = t2;
    t1->right = t3;
    t2->left = t4;
    t4->left = t5;
    t3->left = t6;
    t3->right = t7;
    Solution s = Solution();
    s.mirrorTree(t1);
    print(t1);
    
    return 0;
}

File: 剑指Offer27-er-cha-shu-de-jing-xiang-lcof/Offer27.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def mirrorTree(self, root: TreeNode) -> TreeNode:
        if not root:
            return None
        root.left, root.right = root.right, root.left
        self.mirrorTree(root.left)
        self.mirrorTree(root.right)
        return root

def myprint(t: TreeNode) -> None:
    if not t:
        return None
    q:deque = deque()
    q.append(t)
    print('[', end='')
    while q:
        size = len(q)
        for i in range(size):
            tmp = q.popleft()
            print(tmp.val, end=', ')
            if tmp.left:
                q.append(tmp.left)
            if tmp.right:
                q.append(tmp.right)
    print(']')
if __name__ == '__main__':
    sol = Solution()
    t1 = TreeNode(4)
    t2 = TreeNode(2)
    t3 = TreeNode(7)
    t4 = TreeNode(1)
    t5 = TreeNode(3)
    t6 = TreeNode(6)
    t7 = TreeNode(9)
    t1.left = t2
    t1.right = t3
    t2.left = t4
    t2.right = t5
    t3.left = t6
    t3.right = t7
    sol.mirrorTree(t1)
    myprint(t1)